llama_cpp/_utils.py

```diff
@@ -112,14 +122,59 @@ def get_gpu_info_by_pid(pid) -> float:
         pass
     return 0.0
 
+
 def get_gpu_general_info() -> Tuple[float, float, float]:
     """
     GPU general info (if GPU is available)
     """
     try:
-        gpu_info = subprocess.check_output(["nvidia-smi", "--query-gpu=utilization.gpu,memory.used,memory.free", "--format=csv,noheader"]).decode("utf-8")
-        gpu_utilization, gpu_memory_used, gpu_memory_free = gpu_info.strip().split("\n")[0].split(", ")
-        return tuple(float(tup.split()[0]) for tup in [gpu_utilization, gpu_memory_used, gpu_memory_free])
+        gpu_info = subprocess.check_output(
+            [
+                "nvidia-smi",
+                "--query-gpu=utilization.gpu,memory.used,memory.free",
+                "--format=csv,noheader",
+            ]
+        ).decode("utf-8")
+        gpu_utilization, gpu_memory_used, gpu_memory_free = (
+            gpu_info.strip().split("\n")[0].split(", ")
+        )
+        return tuple(
+            float(tup.split()[0])
+            for tup in [gpu_utilization, gpu_memory_used, gpu_memory_free]
+        )
     except (subprocess.CalledProcessError, FileNotFoundError):
         pass
     return 0.0, 0.0, 0.0
+
+
+async def monitor_task_queue(status_dict: Dict[str, Union[int, float]]):
+    """
+    An asynchronous function that monitors the task queue and updates
+    a shared status dictionary with the number of tasks that have not
+    started and the number of tasks that are currently running.
+    It recursively calls itself to continuously monitor the task queue.
+    NOTE: There will always be 4 tasks running in the task queue:
+    - LifespanOn.main: Main application coroutine
+    - Server.serve: Server coroutine
+    - monitor_task_queue: Task queue monitoring coroutine
+    - RequestReponseCycle.run_asgi: ASGI single cycle coroutine
+    Any upcoming requests will be added to the task queue in the form of
+    another RequestReponseCycle.run_asgi coroutine.
+    """
+    all_tasks = asyncio.all_tasks()
+
+    # Get count of all running tasks
+    _all_tasks = [task for task in all_tasks if task._state == "PENDING"]
+    status_dict["running_tasks_count"] = len(_all_tasks)
+    # Get basic metadata of all running tasks
+    status_dict["running_tasks"] = {
+        task.get_name(): str(task.get_coro())
+        .encode("ascii", errors="ignore")
+        .strip()
+        .decode("ascii")
+        for task in all_tasks
+    }
+
+    asyncio.create_task(
+        monitor_task_queue(status_dict)
```

Comment:
If this portion allows monitor_task_queue to be recursively called, will it print the string output of running_tasks_count and running_tasks every time recursively called, providing the visibility?

Reply:
ah nvm, got it 👍

Reply:
Yep, that's right. Calling the `/health/` endpoint would give such an output, but we are also exporting the `running_tasks` count as a prometheus metric so we can monitor that behaviour with Grafana as well.